modules/puppet_proxy_puppet_api/v3_environmentclassesapi_classes_retriever.rb

```diff
@@ -0,0 +1,100 @@
+    @puppet_api = api || Proxy::PuppetApi::EnvironmentClassesApiv3
+  end
+
+  def classes_in_environment(environment)
+    legacy_classes_format(get_classes(environment))
+  end
+
+  def classes_and_errors_in_environment(environment)
+    classes_and_errors(get_classes(environment))
+  end
+
+  def classes_and_errors(response)
+    response['files'].map do |current|
+      classes = (current['classes'] || []).map do |clazz|
+        parameters = clazz['params'].map do |parameter|
+          if parameter['default_source'].is_a?(String) && parameter['default_source'].start_with?('$')
+            to_return = parameter.dup
+            to_return['default_source'] = "${#{parameter['default_source'].slice(1..-1)}}"
+            to_return
+          else
+            parameter
+          end
+        end
+        {'name' => clazz['name'], 'params' => parameters}
+      end
+      current.has_key?('error') ? current : {'path' => current['path'], 'classes' => classes}
+    end
+  end
+
+  def legacy_classes_format(response)
+    response['files'].map do |current|
+      (current['classes'] || []).map do |manifest| # current['classes'] is nil for 'error' entities
+        parameters = manifest['params'].inject({}) do |all, parameter|
+          parameter_value = parameter['default_literal'] || parameter['default_source']
+          all[parameter['name']] = (parameter_value.is_a?(String) && parameter_value.start_with?('$')) ? "${#{parameter_value.slice(1..-1)}}" : parameter_value
+          all
+        end
+        ::Proxy::Puppet::PuppetClass.new(manifest['name'], parameters)
+      end
+    end.flatten
+  end
+
+  def get_classes(environment)
+    future = async_get_classes(environment)
+    future.value!(MAX_CLIENT_TIMEOUT)
+
+    raise ::Proxy::Puppet::TimeoutError, "Puppet is taking too long to respond, please try again later." if future.pending?
+    future.value
+  end
+
+  def async_get_classes(environment)
+    Concurrent::Future.execute do
+      etag, classes = @m.synchronize { [@etag_cache[environment], @classes_cache[environment]] }
+      begin
+        response, etag = @puppet_api.new(puppet_url, ssl_ca, ssl_cert, ssl_key).list_classes(environment, etag, MAX_PUPPETAPI_TIMEOUT)
+      rescue Exception => e
+        logger.error("Error while retrieving puppet classes for '%s' environment: %s" % [environment, e])
+        raise e
+      end
+
+      if response == Proxy::PuppetApi::EnvironmentClassesApiv3::NOT_MODIFIED
+        classes
+      else
+        @m.synchronize do
+          @etag_cache[environment] = etag
+          @classes_cache[environment] = response
+        end
+      end
+    end
+  end
+end
+
+class Proxy::PuppetApi::EnvironmentClassesCacheInitializer
+  def initialize(classes_retriever, environments_retriever)
+    @environments = environments_retriever
+    @classes = classes_retriever
+  end
+
+  def start_service
+    @environments.all.map(&:name).each {|environment| @classes.get_classes(environment)}
```

Comment:
Secondly, if the smart proxy receives an API request shortly after startup while this preloading is occurring, it looks like there will be multiple requests sent to the Puppet server. Can this be limited so the request is blocked until the previous query has resolved or has timed out?